File: python/problems/leetcode/0094_binary_tree_inorder_traversal/solution.py
"""LeetCode 94: Binary Tree Inorder Traversal."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
        res: list[int] = []
        stack: list[TreeNode] = []
        cur = root
        while cur is not None or stack:
            while cur is not None:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            res.append(cur.val)
            cur = cur.right
        return res


if __name__ == "__main__":
    root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
    assert Solution().inorderTraversal(root) == [1, 3, 2]
    assert Solution().inorderTraversal(None) == []
    assert Solution().inorderTraversal(TreeNode(1)) == [1]
    print("0094 OK")

File: cpp/problems/leetcode/0094_binary_tree_inorder_traversal/solution.cpp
// LeetCode 94
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> st;
        TreeNode* cur = root;
        while (cur || !st.empty()) {
            while (cur) {
                st.push(cur);
                cur = cur->left;
            }
            cur = st.top();
            st.pop();
            res.push_back(cur->val);
            cur = cur->right;
        }
        return res;
    }
};

int main() {
    TreeNode n3(3), n2(2, &n3, nullptr), n1(1, nullptr, &n2);
    assert((Solution().inorderTraversal(&n1) == vector<int>{1, 3, 2}));
    assert(Solution().inorderTraversal(nullptr).empty());
    TreeNode one(1);
    assert((Solution().inorderTraversal(&one) == vector<int>{1}));
    cout << "0094 OK" << endl;
    return 0;
}
